File: chapter22/graph.go
package chapter22

// use to construct graph
type edge struct {
	u int
	v int
	weight int
}

type vertex struct {
	v int
	key int
}

// Vertex represent vertex
type Vertex struct {
	key int
	next *Node
}

// Node represent node
type Node struct {
	v int
	next *Node
}

// Graph represent graph
type Graph struct {
	table []Vertex
	len int
}

type Matrix struct {
	vs []int
	matrix [][]int
	len int
}

// InitGraph init a graph, the vertex should be 1 to n
func InitGraph(vs []vertex, es []edge, n int) Graph {
	graph := Graph{len: n+1}
	graph.table = make([]Vertex, n+1)
	for _, v := range vs {
		graph.table[v.v].key = v.key
	}
	for _, e := range es {
		node := Node{v: e.v}
		node.next = graph.table[e.u].next
		graph.table[e.u].next = &node
	}
	return graph
}

func InitMatrix(vs []vertex, es []edge, n int) Matrix {
	m := Matrix{len: n}
	m.vs = make([]int, n+1)
	for _, v := range vs {
		m.vs[v.v] = v.key
	}
	m.matrix = make([][]int, n+1)
	for i:= 0; i < n + 1; i++ {
		m.matrix[i] = make([]int, n+1)
	}
	for _, e := range es {
		m.matrix[e.u][e.v] = e.weight
	}
	return m
}

File: chapter22/graph.py

from collections import defaultdict


class Vertex:

    def __init__(self, value):
        self.value = value

    def __str__(self):
        return f"{self.value}"

    def __repr__(self):
        return self.__str__()
    
    def __eq__(self, value):
        return self.value == value


class Graph:
    def __init__(self, vertex, edges):
        self.V = {num: Vertex(num) for num in vertex}
        self.Adj = defaultdict(list)
        self.E = edges
        self.W = {}
        for edge in edges:
            u, v, *w = edge
            self.Adj[u].append(v)
            self.Adj[v].append(u)
            if w != []:
                self.W[u, v] = w[0]
                self.W[v, u] = w[0]

    def __str__(self):
        return f"{self.V}, {self.E}"


def initGraph(vertex, edges):
    return Graph(vertex, edges)


def main():
    vertex = [0, 1, 2, 3, 4, 5, 6, 7]
    edges = [(0, 1), (1, 2), (2, 3), (3, 4), (3, 5),
             (4, 5), (4, 7), (4, 6), (5, 6), (6, 7)]
    graph = initGraph(vertex, edges)
    print(graph)


if __name__ == "__main__":
    main()
